src/Compilers/CSharp/Portable/Symbols/MethodSymbol.cs

```diff
@@ -595,6 +595,29 @@ internal bool IsEntryPointCandidate
             get { return IsStatic && Name == WellKnownMemberNames.EntryPointMethodName; }
         }
 
+        internal bool HasAsyncMainReturnType(CSharpCompilation compilation)
+        {
+            var namedType = ReturnType as NamedTypeSymbol;
+            if ((object)namedType == null)
+            {
+                return false;
+            }
+            else if (namedType.ConstructedFrom == compilation.GetWellKnownType(WellKnownType.System_Threading_Tasks_Task))
+            {
+                // Change this to `namedType.IsNonGenericTaskType` if you want to support "task-like" objects.
+                return true;
+            }
+            else if (namedType.ConstructedFrom == compilation.GetWellKnownType(WellKnownType.System_Threading_Tasks_Task_T))
+            {
+                // Change this to `namedType.IsGenericTaskType` if you want to support "task-like" objects.
```

Comment:
Right now we are making a conscious decision to ban task-likes. I left the comment in case we wanted to loosen the rules later on.